### PR TITLE
Allow explicit Touch ID prompts

### PR DESCRIPTION
**Pull request by codingllama**
Allow explicit Touch ID prompts to be triggered via Go code, which will be used to guard the (upcoming) credential picker prompt.

To avoid double-prompting users during Touch ID authentication we have to set a grace period in the underlying LAContext and share it between the functions. Note that AuthContextGuard (native) uses the LAContext explicitly, whereas Authenticate (native) uses it through the SecItemCopyMatching query dictionary.

No UX visible changes are made in the PR, despite the fact that prompting occurs a bit earlier.

#13901

### CODE REVIEW

File: lib/auth/touchid/api_darwin.go [hunk 1]
```diff
@@ -100,6 +105,67 @@ func (touchIDImpl) Diag() (*DiagResult, error) {
 	}, nil
 }
 
+//export runGoFunc
+func runGoFunc(handle C.uintptr_t) {
```
Comment (espadolini) [inline lib/auth/touchid/api_darwin.go:109]:
```suggestion
//export runGoFuncHandle
func runGoFuncHandle(handle C.uintptr_t) {
```

File: lib/auth/touchid/api_darwin.go [hunk 2]
```diff
@@ -50,6 +51,10 @@ const (
 	// rpID are domain names, so it's safe to assume they won't have spaces in them.
 	// https://www.w3.org/TR/webauthn-2/#relying-party-identifier
 	labelSeparator = " "
+
+	// promptReason is the LAContext / Touch ID prompt.
+	// The final prompt is: "$binary is trying to authenticate user".
+	promptReason = "authenticate user"
```
Comment (espadolini) [inline lib/auth/touchid/api_darwin.go:57]: nit: "authenticate the user" perhaps?

File: lib/auth/touchid/context.m [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+#import <LocalAuthentication/LocalAuthentication.h>
+
+#include <stdint.h>
+
+#include <dispatch/dispatch.h>
+
+#include "common.h"
+
+// runGoFunc is provided via CGO exports.
+extern void runGoFunc(uintptr_t handle);
```
Comment (espadolini) [inline lib/auth/touchid/context.m:31]: Shouldn't we include `_cgo_export.h` instead?

File: lib/auth/touchid/context.m [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  }
+
+  return res;
+}
+
+void AuthContextClose(AuthContext *actx) {
+  if (actx == NULL) {
+    return;
+  }
+  actx->la_ctx = NULL;
```
Comment (espadolini) [inline lib/auth/touchid/context.m:83]: Does this take care of deallocating the previous `LAContext`?

File: lib/auth/touchid/api_darwin.go [hunk 5]
```diff
@@ -100,6 +105,67 @@ func (touchIDImpl) Diag() (*DiagResult, error) {
...
+	fn()
+}
+
+// touchIDContext wraps C.AuthContext into an authContext shell.
+type touchIDContext struct {
+	ctx *C.AuthContext
+}
+
+func (c *touchIDContext) Guard(fn func()) error {
+	authCtx := getNativeContext(c)
```
Comment (espadolini) [inline lib/auth/touchid/api_darwin.go:125]: Isn't this just `c.ctx`?

File: lib/auth/touchid/api_darwin.go [hunk 6]
```diff
@@ -100,6 +105,67 @@ func (touchIDImpl) Diag() (*DiagResult, error) {
...
+
+	reasonC := C.CString(promptReason)
+	defer C.free(unsafe.Pointer(reasonC))
+
+	// Passing Go function pointers directly to CGO is not doable, so we pass a
+	// handle and have an exported Go function run it.
+	// See https://github.com/golang/go/wiki/cgo#function-variables.
+	handle := cgo.NewHandle(fn)
```
Comment (espadolini) [inline lib/auth/touchid/api_darwin.go:133]:
```suggestion
	handle := cgo.NewHandle(fn)
	defer handle.Delete()
```

File: lib/auth/touchid/api_darwin.go (hunk 1 above)
Comment (codingllama) [reply lib/auth/touchid/api_darwin.go:109]: Done.

File: lib/auth/touchid/api_darwin.go (hunk 2 above)
Comment (codingllama) [reply lib/auth/touchid/api_darwin.go:57]: I copied the Apple default prompt, which says exactly that in my machine. (YMMV due to localization.)

File: lib/auth/touchid/api_darwin.go (hunk 5 above)
Comment (codingllama) [reply lib/auth/touchid/api_darwin.go:125]: Done. Well spotted.

File: lib/auth/touchid/api_darwin.go (hunk 6 above)
Comment (codingllama) [reply lib/auth/touchid/api_darwin.go:133]: Done, thanks.

**Comment by codingllama** [comment]
Thanks for the quick review, @espadolini. PTAL?

**Review by espadolini — APPROVED** [review]

**Comment by codingllama** [comment]
Friendly ping @capnspacehook ?

**Review by capnspacehook — APPROVED** [review]

**Pull request merged**